pkg/apis/crd/v1beta1/types.go

```diff
@@ -810,3 +810,259 @@ type TLSProtocol struct {
+	ICMPProtocolNumber: "ICMP",
+	IGMPProtocolNumber: "IGMP",
+	SCTPProtocolNumber: "SCTP",
+}
+
+// List the supported destination types in traceflow.
+const (
+	DstTypePod     = "Pod"
+	DstTypeService = "Service"
+	DstTypeIPv4    = "IPv4"
+)
+
+var SupportedDestinationTypes = []string{
+	DstTypePod,
+	DstTypeService,
+	DstTypeIPv4,
+}
+
+// Default timeout in seconds.
+const DefaultTraceflowTimeout int32 = 20
+
+// +genclient
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+type Traceflow struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   TraceflowSpec   `json:"spec,omitempty"`
+	Status TraceflowStatus `json:"status,omitempty"`
+}
+
+// TraceflowSpec describes the spec of the traceflow.
+type TraceflowSpec struct {
+	Source      Source      `json:"source,omitempty"`
+	Destination Destination `json:"destination,omitempty"`
+	Packet      Packet      `json:"packet,omitempty"`
+	// LiveTraffic indicates the Traceflow is to trace the live traffic
+	// rather than an injected packet, when set to true. The first packet of
+	// the first connection that matches the packet spec will be traced.
+	LiveTraffic bool `json:"liveTraffic,omitempty"`
+	// DroppedOnly indicates only the dropped packet should be captured in a
+	// live-traffic Traceflow.
+	DroppedOnly bool `json:"droppedOnly,omitempty"`
+	// Timeout specifies the timeout of the Traceflow in seconds. Defaults
+	// to 20 seconds if not set.
+	Timeout int32 `json:"timeout,omitempty"`
+}
+
+// Source describes the source spec of the traceflow.
+type Source struct {
+	// Namespace is the source namespace.
+	Namespace string `json:"namespace,omitempty"`
+	// Pod is the source pod.
+	Pod string `json:"pod,omitempty"`
+	// IP is the source IPv4 or IPv6 address. IP as the source is supported
+	// only for live-traffic Traceflow.
+	IP string `json:"ip,omitempty"`
+}
+
+// Destination describes the destination spec of the traceflow.
+type Destination struct {
+	// Namespace is the destination namespace.
+	Namespace string `json:"namespace,omitempty"`
+	// Pod is the destination pod, exclusive with destination service.
+	Pod string `json:"pod,omitempty"`
+	// Service is the destination service, exclusive with destination pod.
+	Service string `json:"service,omitempty"`
+	// IP is the destination IPv4 or IPv6 address.
+	IP string `json:"ip,omitempty"`
+}
+
+// IPHeader describes spec of an IPv4 header.
+type IPHeader struct {
+	// Protocol is the IP protocol.
+	Protocol int32 `json:"protocol,omitempty" yaml:"protocol,omitempty"`
+	// TTL is the IP TTL.
+	TTL int32 `json:"ttl,omitempty" yaml:"ttl,omitempty"`
+	// Flags is the flags for IP.
+	Flags int32 `json:"flags,omitempty" yaml:"flags,omitempty"`
+}
+
+// IPv6Header describes spec of an IPv6 header.
+type IPv6Header struct {
+	// NextHeader is the IPv6 protocol.
+	NextHeader *int32 `json:"nextHeader,omitempty" yaml:"nextHeader,omitempty"`
```

Comment:
quick question: why do we use a pointer for `NextHeader`, but not for `Protocol`?
sorry if this was discussed before

Reply:
This is not discussed before for this field, I don't know if there is any specific reason to use pointer for `NextHeader` but not for `Protocol`, @gran-vmv may know the history. 
There was a few discussions about other fields as well, I feel we may go through the whole spec to make the fields type consistent if it's more proper to use 'pointer' in a separate PR with more discussions and verification.